src/Common/src/Microsoft/Win32/SafeHandles/GssSafeHandles.cs

```diff
@@ -0,0 +1,149 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Microsoft.Win32.SafeHandles
+{
+    /// <summary>
+    /// Wrapper around a gss_name_t_desc*
+    /// </summary>
+    internal sealed class SafeGssNameHandle : SafeHandle
+    {
+        public static SafeGssNameHandle CreateUser(string name)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(name), "Invalid user name passed to SafeGssNameHandle create");
+            SafeGssNameHandle retHandle;
+            Interop.NetSecurityNative.Status minorStatus;
+            Interop.NetSecurityNative.Status status = Interop.NetSecurityNative.ImportUserName(
+                out minorStatus, name, Encoding.UTF8.GetByteCount(name), out retHandle);
+
+            if (status != Interop.NetSecurityNative.Status.GSS_S_COMPLETE)
+            {
+                retHandle.Dispose();
+                throw new Interop.NetSecurityNative.GssApiException(status, minorStatus);
+            }
+
+            return retHandle;
+        }
+
+        public static SafeGssNameHandle CreatePrincipal(string name)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(name), "Invalid principal passed to SafeGssNameHandle create");
+            SafeGssNameHandle retHandle;
+            Interop.NetSecurityNative.Status minorStatus;
+            Interop.NetSecurityNative.Status status = Interop.NetSecurityNative.ImportPrincipalName(
+                out minorStatus, name, Encoding.UTF8.GetByteCount(name), out retHandle);
+
+            if (status != Interop.NetSecurityNative.Status.GSS_S_COMPLETE)
+            {
+                retHandle.Dispose();
+                throw new Interop.NetSecurityNative.GssApiException(status, minorStatus);
+            }
+
+            return retHandle;
+        }
+
+        public override bool IsInvalid
+        {
+            get { return handle == IntPtr.Zero; }
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            Interop.NetSecurityNative.Status minorStatus;
+            Interop.NetSecurityNative.Status status = Interop.NetSecurityNative.ReleaseName(out minorStatus, ref handle);
+            SetHandle(IntPtr.Zero);
+            return status == Interop.NetSecurityNative.Status.GSS_S_COMPLETE;
+        }
+
+        private SafeGssNameHandle()
+            : base(IntPtr.Zero, true)
+        {
+        }
+    }
+
+    /// <summary>
+    /// Wrapper around a gss_cred_id_t_desc_struct*
+    /// </summary>
+    internal class SafeGssCredHandle : SafeHandle
+    {
+        /// <summary>
+        ///  returns the handle for the given credentials.
+        ///  The method returns null if the username is null or empty.
```

Comment:
The method doesn't really return `null`, it returns a wrapped null.  "returns null" suggests the caller needs to be wary when marshalling or dereferencing.